Vislib Point Series Charts

ppisljar opened the pull request:
*Summary*:


In Kibana Visualize, the ability for users to customize and style charts has been greatly improved. Users can now overlay multiple chart types on a single plot, use horizontal layouts, and modify the styling of axes and gridlines.

![image](https://cloud.githubusercontent.com/assets/1833023/25592217/ecc8472a-2e85-11e7-9c35-bd02be93e080.png)



---------


Vislib Point Series Charts - adds new functionality to Axis, Line and Bar charts

Multiple chart types per plot #4566
Add option to suppress scale values #9026
Second Y-Axis #2390
Horizontal bar chart #1396
Optional chart grids #5747

epixa commented:
Let's scale up test coverage on this.  A 2000 line PR will likely have a considerable amount of test changes/additions.

epixa commented:
Is this PR reviewable at this point, or are you just putting it here as a WIP?

ppisljar commented:
just WIP

- [ ]  selenium tests
- [ ] update show check boxes to match the one in heatmap chart
- [ ] stepped line
- [ ] update docs

jimmyjones2 commented:
Is it possible to address #2021 in this PR as well?

ppisljar commented:
i would prefer to keep this PR as small as possible (its already big). once this is merged adding things like #2021 will be quite easy.

ppisljar commented:
I updated based on suggestions I got in the UI review:

thomasneirynck commented:
jenkins, test this

thomasneirynck commented:
needs a merge/rebase

thomasneirynck commented:
Below comments are for area-charts in particular, not the smaller changes to the other charts.

---

I think the splitting the options in tabs is a good idea. It avoids overwhelming the user. It is a great improvement over the earlier design with everything in a single tab.

I really like the power of the charts. It solves many pain points of current users, notably overlapping multiple chart types and offering a more data-driven design.
 
After working with the new Area-charts for a while, I am concerned about the current naming and UI, which remains complex.  It is hard to work with. I think partially due to:

- We introduce unfamiliar terminology like `Value-axis` and `Category-axis`.
- We use "generic"-sounding terms like 'Series', 'Dot size' for very specific things
- The four tabs are interdependent. I see how the Data-tab informs the other tabs, but we are making modifications to the value-axes in both Axes, Series, and Other Options.

---

With the risk of this being just another opinion to throw on the pile, I propose a different scheme of organizing the UI, without taking away any of the power.

1. We introduced new names for things that already have names. We are creating common `f(x)=y` charts. IMHO, there is a more familiar vocabulary already in place for these charts. 
- `X` is the  independent variable, `
- `Y` is the dependent variable, or the measurement for a given feature X.

In the new scheme, we should replace all occurences as follows:
`Value-axis` -> `Y-axis`
`Category-axis` -> `X-axis`

2. I don't think we need to worry too much about how `XY` would imply the actual rotation of the chart. 
Right now, we need to define this in the 'Axes' tab at the `Category Axes`-section. This selection then cascades down to the `Value Axes`. 

This whole flow I would just change to a single combo-box: `[horizontal layout|vertical layout]`. For the time being, I would omit the option of having the `category-axes` be either right or top aligned.

3.  `Series` does not seem like a good name to indicate the chart-type. Also, these are in fact options for the `Value Axes`, so it seems it is somewhat floating apart. I would do away with this section, and move these options to the existing modifiers in the `Axes` tab.

4. `Dot size` is odd name. Not sure what to do here ;)

5. I am not sure if we need to be as fine-grained to allow users to manually configure each value-axis. Why can't we just correspond to a metric to a Value-axes?

6. allowing any number of value-axes makes the UI perhaps a little too generic. I feel limiting this to 2 would focus the chart's purpose more, and keep the UI more simple.

Given these thoughts, I think we should consider following rearrangement:

We retain the four tabs, but rearrange some of the contents:
1. `Data`
- the same as now 
- (but with something better than 'dot-size')
2. `X-axis`
- this contains now the options of `Axes-Category Axis`
- remove the `Position` combo-box
- add a show-gridlines checkbox
3. `Y-axes`
For each Y-axes (corresponding to a metric in Data)
- This contains the options of `Axes>ValueAxis*`
- This also contains the options of what is now in the `Series`-tab.
- remove the `Position` combo-box
- use `Chart Type` instead of `Series` to indicate options like type & mode.
- add a show-gridlines checkbox
4. `Options`
- keep `Other options>Settings`
- remove `Other options->Grid`
- add a 'horizontal' or 'vertical' layout combo-box. This would flip the chart. As for label position, let's assume label positions are relative to the axis-line, not the chart (then they can rotate along).

---

All the above is somewhat speculative, and I don't expect it to be the best solution.

The gist I think is this. This is an awesome introduction of advanced charting, but very dense in terms of new functionality.

Let's consider:
- make this more simple: mainly, imho, this is restricting the number of `value-axis`.
- try and use more commonplace naming.
- avoid as many dependencies between the different tabs.

ppisljar commented:
Im gonna try to answer to your comments above one by one .... let me know if i missed anything:

- Value Axis/ Category Axis terminology ... seems we have a lot of different opinions around this one. I am gonna change it back to X/Y axis (thats how it was before UI review and before this PR) and lets stick with that. If we decide to change that it should be a separate PR.

- 'Dot Size' ... nothing has changed, it was like this before this PR. i would prefer to keep this PR centered on what it is and do all the not-directly-changed-by-this things out of this scope and handled in separate PR.

- yes, the four tabs are interdependent ... after all they are all options for one chart. however I don't agree that series and other settings tabs are modifying axes ... ? which option do you think it does that ?
 - for series ... you can choose the value axis ... but thats it. you cant change it
 - for other options .... you can set the grid lines on value axis ... but you cant change the value axis


2. X/Y naming ...  i don't know ... i would associate X with horizontal usually. But as proposed above, lets not change the naming (so lets stick to the X/Y we had before this PR) ... and lets update the naming in separate PR if we decide to do so.

3. Series is what is used in every other charting tool I saw, but i agree it doesn't apply best here but i have no suggestion for a better term. there are no options for the value axes. You only select WHICH value axis a series is using. think of it as foreign key  .... we cant move this settings to the axes ... (well we could move it under the same tab .... but then splitting tabs didn't make that much sense (we have just the grid in the last tab ... and this is actually the tab that would grow the most (adding multiple series) so i like it being separate.

4. Dot size .... this PR didn't touch it.

5.
 - we don't want every metric on separate axis.
 - axis options are nothing new ... only now with multiple axes you have them defined per axis. i think its a must have if you have multiple axes (multiple axes ... but all with same config ... sounds wrong)
 - we definetely don't want to stick that options to Data tab. lets keep that about data selection only.

6. i am sure we need more than 2. i am also sure we don't need unlimited. however my feeling about this is that this are advanced options, so lets give the responsibility to use them correctly to the user. if he wants 10 axes ... im ok with it. If we would to imply a limit i would go with 4 and nothing less.

7. 
- grid lines: i don't think it would be better this way. so right now you have one place to set your grid. you would split it up to two places. (two tabs actually) which i think will be confusing for the user. also it will allow them to add multiple value axis grid lines which is not very usefull.
I think that when you want grid on your chart ... you look for grid ... and you are not thinking about grid when you are configuring your x/y axes.

- we cant remove position, its already there ... so that would be a regression. 



------- finally:

making more simple: i dont think restricting of value axis will make it simpler. 
- whole options i would consider advanced setting, simple user would not go there ...
- so if only an advanced user would go there ... i think its valid to expect he might need more than 2 axes
- also interface which allows you to add only one would not be any simpler i think ?

use better naming: lets go back to X/Y axis .... lets not change the naming of things that were not directly changed by this PR (like dot size) ... and lets try to improve the naming of the new options this PR added.

dependencies between tabs: as little as possible. however i series specifying value axes in a series tab ... i woulnd't count that as a dependency. you have value axes, you have series, they are connected ... but separate.

ppisljar commented:
@thomasneirynck  i hope above comment doesn't sound like "no i am gonna keep it this way". I really liked your long comment, it had me rethink some things which is really good. I am also open for discussing anything about you think we should further discuss. But I do have some strong opinions on this as I already put quite a lot of taught into this  ... please bear with me :)

ppisljar commented:
yesterday we had a zoom session with CJ and we came up with these changes (already implemented)

- reverse axis and series tab (series should come first)
- reverse value axis and cat axis (to match order we have on Data tab)
- new axis in dropdown takes you to the axis tab (so you dont need to go back and forth)
- instead of other options rename to settings
- expand grid options by default

ppisljar commented:
PS: selenium tests are failing due to rearrangement of UI .... i'm gonna wait a bit in case we do some more reorganization before fixing them.

thomasneirynck commented:
@ppisljar that's alright, I'm not wedded to any of the suggestions ;)!

kobelb commented:
Nice work! I'm really impressed with the level of configuration that you've exposed to the user, and the unification of the three chart types is awesome. 

A lot has changed since the last UI review, and I feel like it'd be beneficial to present what's currently implemented to a larger audience. Could we potentially host this version like you did for the Heatmap demo?

There are a few UI inconsistencies between the tabs that could be addressed to increase usability. However, I also have some hesitation regarding the experience of configuring the Axis, but I might be alone in this opinion, hence my recommendation that we demo this to a larger audience.

ppisljar commented:
https://ppisljar.hopto.org .... its still there

I got some feedback from the previous version of demo running there, but apart from some bugs ppl noticed nothing concrete except that they liked the new features. We probably should ask more concrete questions if we are looking to get more concrete answers.

ppisljar commented:
We had another meeting about UI for this today and this is the list of things we came up with:

- series -- 'show' option is not necessary
- visually split sections under one tab better
- axis naming: default name should tell you about axis position (LeftAxis-1, RightAxis-2) ... if title is added to the axis it should be called just that.
- second axis should be positioned on the right by default
- combine series + axis under 'Advanced' tab

ppisljar commented:
I had a live usability test with Loek van Gool today, you can view the recording [here](https://drive.google.com/a/elastic.co/file/d/0B5MC_aCRfV80cEZ2a1dyNHJBWlE/view?usp=sharing) 

after that I settled on two tabs ('Axes & Metrics' and 'Panel Options'). 

@kobelb @thomasneirynck can you start with code review ? in the meantime I will try to get some more SAs to test it out.

kobelb commented on src/core_plugins/kbn_vislib_vis_types/public/horizontal_bar.js:
```diff
@@ -0,0 +1,152 @@
+import VisSchemasProvider from 'ui/vis/schemas';
+import pointSeriesTemplate from 'plugins/kbn_vislib_vis_types/editors/point_series.html';
+
+export default function PointSeriesVisType(Private) {
```
I'm getting the following console error when trying to create a horizontal bar chart
```
TypeError: visType is not a function
    at new VisConfig (vis_config.js:50)
    at Vis.render (vis.js:122)
    at Vis.resize (vis.js:153)
    at Function.wrapper (index.js:3095)
    at events.js:60
    at processQueue (angular.js:14745)
    at angular.js:14761
    at Scope.$eval (angular.js:15989)
    at Scope.$digest (angular.js:15800)
    at angular.js:16028
```

kobelb commented on src/core_plugins/kbn_vislib_vis_types/public/histogram.js:
```diff
@@ -1,8 +1,8 @@
 import VislibVisTypeVislibVisTypeProvider from 'ui/vislib_vis_type/vislib_vis_type';
 import VisSchemasProvider from 'ui/vis/schemas';
-import histogramTemplate from 'plugins/kbn_vislib_vis_types/editors/histogram.html';
```
Can we get rid of `histogram.html` now?

kobelb commented on src/core_plugins/kbn_vislib_vis_types/public/controls/point_series/series.js:
```diff
@@ -0,0 +1,80 @@
+          }
+          serieCount++;
+        });
+        $scope.vis.params.seriesParams = $scope.vis.params.seriesParams.slice(0, serieCount);
```
I'm afraid I'm not seeing why we're doing this here.

ppisljar replied:
i guess i need to write it in a better way.

seriesParams could have 10 series in it ... when you remove 2 of the series, 8 of the seriesParams will be updated and the rest will be thrown away.

kobelb commented on src/core_plugins/kbn_vislib_vis_types/public/controls/point_series/series.js:
```diff
@@ -0,0 +1,80 @@
+        return $scope.vis.params.seriesParams.map(series => series.type).join();
+      }, () => {
+        const types = _.uniq(_.map($scope.vis.params.seriesParams, 'type'));
+        $scope.savedVis.type = types.length === 1 ? types[0] : 'histogram';
```
Would you mind explaining what's going on here?

ppisljar replied:
when we are saving the vis we want to update the type (so if you started with line chart and you then changed it to bar chart we want bar chart icon to appear)

i am not sure this condition is necessary but if you would to save a chart with nothing on it ... it would save as histogram.

ppisljar commented on src/core_plugins/kbn_vislib_vis_types/public/controls/point_series/series.js:
```diff
@@ -0,0 +1,80 @@
+          try {
+            return agg.makeLabel();
+          } catch (e) {
+            return '';
```
this happens when you select an aggregation and your field is not yet selected (makeLabel tries to access the field name)

ppisljar commented on src/ui/public/vislib/lib/types/point_series.js:
```diff
@@ -3,25 +3,49 @@ import errors from 'ui/errors';
-    const stacked = ['stacked', 'percentage', 'wiggle', 'silhouette'].includes(cfg.mode);
-    let interpolate = cfg.interpolate;
+  const createSerieFromParams = (cfg, seri) => {
+    // todo this wont work with splits ... same issue exists in dispatch
```
actually i need to test this out, i think it might already be fixed

ppisljar replied:
yeah this is already sorted out

kobelb commented:
Really minor nit-pick, but the Horizontal Bar Chart cuts off the x-axis label: 

<img width="562" alt="screen shot 2017-02-10 at 6 28 29 am" src="https://cloud.githubusercontent.com/assets/627123/22825289/530d441a-ef5a-11e6-9045-9c4f998a25c9.png">

thomasneirynck commented:
jenkins, test this

thomasneirynck commented on src/core_plugins/kbn_vislib_vis_types/public/controls/point_series/trend_lines.html:
```diff
@@ -0,0 +1,41 @@
+
+    <div class="vis-editor-agg-header-controls btn-group" style="float: right;">
+      <!-- remove button -->
+      <button aria-label="Add value axis" ng-click="addTrendLine()" tooltip="Add value axis" tooltip-append-to-body="true" type="button" class="btn btn-xs btn-primary ng-scope">
```
consider changing this to "add Y axis"

ppisljar replied:
actually this is not used, so i am gonna remove it

thomasneirynck commented:
there's an uncaught error. 

![image](https://cloud.githubusercontent.com/assets/1833023/22938647/5ec4aa70-f2aa-11e6-9d34-73d9d5418219.png)

To reproduce;

- populate ES with makelogs data
- Open new visualization app in Kibana
- select line chart
- see error in logs

thomasneirynck commented:
The area-chart error pops up in situations that are not entirely meaningful:

![image](https://cloud.githubusercontent.com/assets/1833023/22938803/e382100e-f2aa-11e6-8ceb-b6fd60de068f.png)

In this case, this is just because there's only a single measurement, not because the area chart doesn't have a sub-aggregation.

For example, compared to the line chart, it pops up nicely.

![image](https://cloud.githubusercontent.com/assets/1833023/22938928/348eff48-f2ab-11e6-9ef2-23193287491f.png)

Not sure if the error-message needs to be fixed in the context of this PR. However, since the new UI is so much more flexible, this error might pop-up more frequently than earlier.

I believe this probably has a similar root issue that triggers the orphaned popup in #10324.

thomasneirynck commented:
How do I get one metric on the left-axis (=average machine.ram), and the other metric (count) on the right-axis? I feel like I'm doing something obviously wrong ...

![image](https://cloud.githubusercontent.com/assets/1833023/22939364/6ab9252a-f2ac-11e6-92e0-93b1bd40e083.png)

thomasneirynck commented on src/core_plugins/kbn_vislib_vis_types/public/histogram.js:
```diff
@@ -14,46 +14,122 @@ export default function HistogramVisType(Private) {
+      optionTabs: [
+        {
+          name: 'advanced',
+          title: 'Axes & Metrics',
```
consider Metrics & Axis, given the first section is metrics, and the second is axes

thomasneirynck commented on docs/visualize/xychart.asciidoc:
```diff
@@ -0,0 +1,126 @@
+*Position*:: You can choose where you want to display category axis. If you position your category axis on the left or right the chart will turn to the horizontal type.
+*Advanced Options*::
+*Labels - Show Labels*:::: Allows you to hide axis labels
+*Labels - Filter Labels*:::: If filter labels is enabled some labels will be hidden in case there is not enough spave to display them
```
typo: 'space' iso 'spave'

thomasneirynck commented on docs/visualize/xychart.asciidoc:
```diff
@@ -0,0 +1,126 @@
+
+*Show*:: You can decide to hide the value axis completely
+*Label*:: Allows to define a custom label
+*Position*:: Options for position depend on the position of your category axis. If category axis is positioned on the top or bottom you can position value axis on the left or right. In the oposite case you can position your value axis on the top or bottom.
```
typo: `opposite` iso `oposite`

thomasneirynck commented on src/core_plugins/kbn_vislib_vis_types/public/controls/point_series/category_axis.js:
```diff
@@ -0,0 +1,33 @@
+import vislibValueAxesTemplate from 'plugins/kbn_vislib_vis_types/controls/point_series/category_axis.html';
+const module = uiModules.get('kibana');
+
+module.directive('vislibCategoryAxis', function ($parse, $compile) {
```
consider removing unused imports $parse, $compile. Similar to other point_series/* directives.

thomasneirynck commented on src/core_plugins/kbn_vislib_vis_types/public/controls/point_series/category_axis.js:
```diff
@@ -0,0 +1,33 @@
+import _ from 'lodash';
```
remove unused _ imports. similar to other files here. jQuery $-imports are generally unused too.

ppisljar commented:
thanks @thomasneirynck 

I updated based on your comments and fixed the error. I agree that area chart error should be handled in separate PR.

this is ready for another look.

thomasneirynck commented:
jenkins, test this

tbragin commented:
Some examples of charts that this pull enables:

<img width="1399" alt="horizontal_bar_chart" src="https://cloud.githubusercontent.com/assets/933397/23066683/b3b19958-f4d0-11e6-8112-c6875a11440d.png">

<img width="1399" alt="combination_chart" src="https://cloud.githubusercontent.com/assets/933397/23066685/b69c0e82-f4d0-11e6-80ff-75fa2b4c306f.png">

ndimer commented:
Hey Guys! When these changes will be in the GA version of Kibana? Is there any released version of Kibana with secondary y-axis support?

Thanks!

ppisljar commented:
unfortunately not. if everything goes well this will be part of 5.4 ... lets keep our fingers crossed :)

ndimer commented:
Thanks Peter! Any estimated release date?

> On 20 Feb 2017, at 20:53, Peter Pisljar <notifications@github.com> wrote:
> 
> unfortunately not. if everything goes well this will be part of 5.4 ... lets keep our fingers crossed :)
> 
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

epixa commented:
@ndimer Sorry, but we don't put estimates on release dates since they can change over time based on circumstances and features.

pmriyazu commented:
Hi
In which version dual Y axis is available.otherwise can u tell me the changes to make in our current version. bcoz I have demo in two days and I need to implement that. could you plz provide me the changes.

ppisljar commented:
@pmriyazu its sufficient to ask the question on one thread. 
as i replied in the others, this will be available in 5.x if everything goes well.
